edlib/src/edlib.tpp

```diff
@@ -78,85 +90,95 @@ class EqualityDefinition {
-                                           int* bestScore_, int** positions_, int* numPositions_);
-
-static int myersCalcEditDistanceNW(const Word* Peq, int W, int maxNumBlocks,
-                                   int queryLength,
-                                   const unsigned char* target, int targetLength,
-                                   int k, int* bestScore_,
-                                   int* position_, bool findAlignment,
-                                   AlignmentData** alignData, int targetStopPosition);
-
-
-static int obtainAlignment(
-        const unsigned char* query, const unsigned char* rQuery, int queryLength,
-        const unsigned char* target, const unsigned char* rTarget, int targetLength,
-        const EqualityDefinition& equalityDefinition, int alphabetLength, int bestScore,
+template<class AlphabetIdx>
+int myersCalcEditDistanceSemiGlobal(const Word* Peq, int W, int maxNumBlocks,
+                                    int queryLength,
+                                    const AlphabetIdx* target, int targetLength,
+                                    int k, EdlibAlignMode mode,
+                                    int* bestScore_, int** positions_, int* numPositions_);
+
+template<class AlphabetIdx>
+int myersCalcEditDistanceNW(const Word* Peq, int W, int maxNumBlocks,
+                            int queryLength, const AlphabetIdx* target, int targetLength,
+                            int k, int* bestScore_,
+                            int* position_, bool findAlignment,
+                            AlignmentData** alignData, int targetStopPosition);
+
+template<class AlphabetIdx>
+int obtainAlignment(
+        const AlphabetIdx* query, const AlphabetIdx* rQuery, int queryLength,
+        const AlphabetIdx* target, const AlphabetIdx* rTarget, int targetLength,
+        const EqualityDefinition<AlphabetIdx>& equalityDefinition, int alphabetLength, int bestScore,
         unsigned char** alignment, int* alignmentLength);
 
-static int obtainAlignmentHirschberg(
-        const unsigned char* query, const unsigned char* rQuery, int queryLength,
-        const unsigned char* target, const unsigned char* rTarget, int targetLength,
-        const EqualityDefinition& equalityDefinition, int alphabetLength, int bestScore,
+template<class AlphabetIdx>
+int obtainAlignmentHirschberg(
+        const AlphabetIdx* query, const AlphabetIdx* rQuery, int queryLength,
+        const AlphabetIdx* target, const AlphabetIdx* rTarget, int targetLength,
+        const EqualityDefinition<AlphabetIdx>& equalityDefinition, int alphabetLength, int bestScore,
         unsigned char** alignment, int* alignmentLength);
 
-static int obtainAlignmentTraceback(int queryLength, int targetLength,
-                                    int bestScore, const AlignmentData* alignData,
-                                    unsigned char** alignment, int* alignmentLength);
+int obtainAlignmentTraceback(int queryLength, int targetLength,
```

Comment:
Does this function also have to be inlined?

In case that is true, and we missed that, I think it would be good to create a test for this, because we are certainly going to miss it again in the future -> at any moment somebody can add a function, forget to add inline, and somebody somewhere will have an error and we will not know!
How can we test this problem -> I guess it is best that we replicate the scenario that Paolo described and just check that it compiles!